src/lib/operators/sort_new.cpp

```diff
@@ -0,0 +1,301 @@
+
+      value_segment_value_vector.reserve(row_count_out);
+      value_segment_null_vector.reserve(row_count_out);
+
+      auto segment_ptr_and_accessor_by_chunk_id =
+          std::unordered_map<ChunkID, std::pair<std::shared_ptr<const BaseSegment>,
+                                                std::shared_ptr<AbstractSegmentAccessor<ColumnDataType>>>>();
+      segment_ptr_and_accessor_by_chunk_id.reserve(row_count_out);
+
+      for (auto row_index = 0u; row_index < row_count_out; ++row_index) {
+        const auto [chunk_id, chunk_offset] = pos_list->operator[](row_index);
+
+        auto& segment_ptr_and_typed_ptr_pair = segment_ptr_and_accessor_by_chunk_id[chunk_id];
+        auto& base_segment = segment_ptr_and_typed_ptr_pair.first;
+        auto& accessor = segment_ptr_and_typed_ptr_pair.second;
+
+        if (!base_segment) {
+          base_segment = _input_left->get_output()->get_chunk(chunk_id)->get_segment(column_id);
+          accessor = create_segment_accessor<ColumnDataType>(base_segment);
+        }
+
+        // If the input segment is not a ReferenceSegment, we can take a fast(er) path
+        if (accessor) {
+          const auto typed_value = accessor->access(chunk_offset);
+          const auto is_null = !typed_value;
+          value_segment_value_vector.push_back(is_null ? ColumnDataType{} : typed_value.value());
+          value_segment_null_vector.push_back(is_null);
+        } else {
+          const auto value = (*base_segment)[chunk_offset];
+          const auto is_null = variant_is_null(value);
+          value_segment_value_vector.push_back(is_null ? ColumnDataType{} : boost::get<ColumnDataType>(value));
+          value_segment_null_vector.push_back(is_null);
+        }
+
+        ++chunk_offset_out;
+
+        // Check if value segment is full
+        if (chunk_offset_out >= _output_chunk_size) {
+          chunk_offset_out = 0u;
+          auto value_segment = std::make_shared<ValueSegment<ColumnDataType>>(std::move(value_segment_value_vector),
+                                                                              std::move(value_segment_null_vector));
+          chunk_it->push_back(value_segment);
+          value_segment_value_vector = pmr_concurrent_vector<ColumnDataType>();
+          value_segment_null_vector = pmr_concurrent_vector<bool>();
+          ++chunk_it;
+        }
+      }
+
+      // Last segment has not been added
+      if (chunk_offset_out > 0u) {
+        auto value_segment = std::make_shared<ValueSegment<ColumnDataType>>(std::move(value_segment_value_vector),
+                                                                            std::move(value_segment_null_vector));
+        chunk_it->push_back(value_segment);
+      }
+    });
+  }
+
+  for (auto& segments : output_segments_by_chunk) {
+    output->append_chunk(segments);
+  }
+
+  return output;
+}
+
+void SortNew::_on_cleanup() {
+  // TODO(anyone): Implement, if necessary
```

Comment:
As discussed, let's make sure that we follow the hyrise conventions here, i. e. clean-up e. g. any shared ptrs being stored in member variables and being only created during the execution of the operator itself.